benchmarks/experimental_components/utils/fuzzer.py

```diff
@@ -0,0 +1,188 @@
+import functools
+
+import numpy as np
+import torch
+
+
+__all__ = ["Fuzzer", "FuzzedParameter", "FuzzedTensor"]
+
+
+_DISTRIBUTIONS = (
+    "loguniform",
+    "uniform",
+)
+
+class FuzzedParameter(object):
+    def __init__(self, name, minval=None, maxval=None, distribution=None, constraint=None):
+        self._name = name
+        self._minval = minval
+        self._maxval = maxval
+        self._distribution = self._check_distribution(distribution)
+        self._contraint = constraint
+
+    @property
+    def name(self):
+        return self._name
+
+    def sample(self, state):
+        if self._distribution == "loguniform":
+            return self._loguniform(state)
+
+        if self._distribution == "uniform":
+            return self._uniform(state)
+
+        if isinstance(self._distribution, dict):
+            return self._custom_distribution(state)
+
+    def satisfies_constraints(self, params):
+        return self._contraint is None or self._contraint(**params)
+
+    def _check_distribution(self, distribution):
+        if not isinstance(distribution, dict):
+            assert distribution in _DISTRIBUTIONS
+        else:
+            assert abs(sum(distribution.values()) - 1) <= 1e-5, "Distribution is not normalized"
+            assert self._minval is None
+            assert self._maxval is None
+
+        return distribution
+
+    def _loguniform(self, state):
+        output = int(2 ** state.uniform(
+            low=np.log2(self._minval),
+            high=np.log2(self._maxval)
+        ))
+        if output < self._minval:
+            return self._minval
+        if output > self._maxval:
+            return self._maxval
+        return output
+
+    def _uniform(self, state):
+        if isinstance(self._minval, int) and isinstance(self._maxval, int):
+            return int(state.randint(low=self._minval, high=self._maxval + 1))
+        return state.uniform(low=self._minval, high=self._maxval)
+
+    def _custom_distribution(self, state):
+        return state.choice(tuple(self._distribution.keys()), p=tuple(self._distribution.values()))
+
+
+class FuzzedTensor(object):
+    def __init__(self, name, size, probability_contiguous=0.5,
+                 min_elements=None, max_elements=None,
+                 dim_parameter=None, roll_parameter=None,
+                 tensor_constructor=None):
+        self._name = name
+        self._size = size
+        self._probability_contiguous = probability_contiguous
+        self._min_elements = min_elements
+        self._max_elements = max_elements
+        self._dim_parameter = dim_parameter
+        self._roll_parameter = roll_parameter
+        self._tensor_constructor = tensor_constructor
+
+    @property
+    def name(self):
+        return self._name
+
+    def _make_tensor(self, params, state):
+        size = self._get_concrete_size(params)
+        dim = len(size)
+        if self._tensor_constructor is None:
+            tensor = torch.rand(size=size)
```

Comment:
There should be a dtype argument somewhere, and default constructor should take it into account (e.g. `rand` is not supported for integer dtypes). I remember you were already mentioning it

Reply:
I added dtype to FuzzedTensor. (Since a Fuzzer can generate Tensors of different dtypes. E.g. a value Tensor and a mask Tensor.)